docarray/storage/abstract_doc_store.py

```diff
@@ -0,0 +1,646 @@
+        :param search_field: name of the field to search on
+        :param limit: maximum number of documents to return
+        :return: a named tuple containing `documents` and `scores`
+        """
+        if isinstance(queries[0], BaseDocument):
+            query_docs: Sequence[BaseDocument] = cast(Sequence[BaseDocument], queries)
+            query_texts: Sequence[str] = self._get_values_by_column(
+                query_docs, search_field
+            )
+        else:
+            query_texts = cast(Sequence[str], queries)
+        return self._text_search_batched(
+            query_texts, search_field=search_field, limit=limit, **kwargs
+        )
+
+    ##########################################################
+    # Helper methods                                         #
+    # These might be useful in your subclass implementation  #
+    ##########################################################
+
+    @staticmethod
+    def _get_values_by_column(docs: Sequence[BaseDocument], col_name: str) -> List[Any]:
+        """Get the value of a column of a document.
+
+        :param docs: The DocumentArray to get the values from
+        :param col_name: The name of the column, e.g. 'text' or 'image__tensor'
+        :return: The value of the column of `doc`
+        """
+        leaf_vals = []
+        for doc in docs:
+            if '__' in col_name:
+                fields = col_name.split('__')
+                leaf_doc: BaseDocument = doc
+                for f in fields[:-1]:
+                    leaf_doc = getattr(leaf_doc, f)
+                leaf_vals.append(getattr(leaf_doc, fields[-1]))
+            else:
+                leaf_vals.append(getattr(doc, col_name))
+        return leaf_vals
+
+    def _get_col_value_dict(
+        self, docs: Union[BaseDocument, Sequence[BaseDocument]]
+    ) -> Dict[str, Generator[Any, None, None]]:
+        """
+        Get all data from a (sequence of) document(s), flattened out by column.
+        :param docs: The document(s) to get the data from
+        :return: A dictionary mapping column names to a generator of values
+        """
+        if isinstance(docs, BaseDocument):
+            docs_seq: Sequence[BaseDocument] = [docs]
+        else:
+            docs_seq = docs
+        if not self._is_schema_compatible(docs_seq):
+            raise ValueError(
+                'The schema of the documents to be indexed is not compatible'
+                ' with the schema of the store.'
+            )
+
+        def _col_gen(col_name: str):
+            return (self._get_values_by_column([doc], col_name)[0] for doc in docs_seq)
+
+        return {col_name: _col_gen(col_name) for col_name in self._columns}
+
+    ##################################################
+    # Behind-the-scenes magic                        #
+    # Subclasses should not need to implement these  #
+    ##################################################
+    def __class_getitem__(cls, item: Type[TSchema]):
```

Comment:
here you need to do the same magic as in DOcumentArray to handle the case where the item is a string or a typevar

Reply:
maybe at some point we might want to abstract aways this templating behavior. But not now I guess

Reply:
you're right